### PR TITLE
Fix/performance profiling

### PR DESCRIPTION
**Pull request by angelskieglazki**

### CODE REVIEW

File: src/trudp_channel.c [hunk 1]
```diff
@@ -623,6 +631,29 @@ int trudpChannelProcessReceivedPacket(trudpChannelData *tcd, uint8_t *data,
...
+            while((el = teoMapIteratorNext(&it))) {
+                trudpChannelData *tcd = (trudpChannelData *)teoMapIteratorElementData(el, NULL);
+                uint32_t timeout_sq = trudpChannelSendQueueGetTimeout(tcd, current_time);
+                if(timeout_sq < min_timeout_sq) {
+                  min_timeout_sq = timeout_sq;
+                  tcd->td->expected_max_time = timeout_sq;
+                  tcd->td->channel_key = tcd->channel_key;
+                }
+                if(!min_timeout_sq) break;
+            }
```
Comment (chernikovdmitry) [inline src/trudp_channel.c]: Code like this exist in two places - here and in trudp.c:189, would much better to extract it to function

File: src/trudp.c [hunk 2]
```diff
@@ -182,6 +185,23 @@ void trudpChannelDestroyAddr(trudpData *td, const char *addr, int port, int chan
...
+                while((el = teoMapIteratorNext(&it))) {
+                    trudpChannelData *tcd = (trudpChannelData *)teoMapIteratorElementData(el, NULL);
+                    uint32_t timeout_sq = trudpChannelSendQueueGetTimeout(tcd, current_time);
+                    if(timeout_sq < min_timeout_sq) {
+                    min_timeout_sq = timeout_sq;
+                    tcd->td->expected_max_time = timeout_sq;
+                    tcd->td->channel_key = tcd->channel_key;
+                    }
+                    if(!min_timeout_sq) break;
+                }
```
Comment (chernikovdmitry) [inline src/trudp.c]: Code like this exist in two places - here and in trudp_channel.c:635, would much better to extract it to function

**Review by chernikovdmitry — CHANGES_REQUESTED** [review]
Extract some identical code to a function, please

File: src/trudp_channel.c [hunk 3]
```diff
@@ -623,6 +630,40 @@ int trudpChannelProcessReceivedPacket(trudpChannelData *tcd, uint8_t *data,
...
+                  min_time = current_time;
+                  min_channel_key = tcd->channel_key;
+                  break;
+                } else if (expected_time < min_time) {
+                  min_time = expected_time;
+                  min_channel_key = tcd->channel_key;
+                }
+            }
+            tcd->td->expected_max_time = min_time;
+            tcd->td->channel_key = min_channel_key;
```
Comment (chernikovdmitry) [inline src/trudp_channel.c]: trudpRecalculateExpectedSendTime(tcd->td);

**Review by chernikovdmitry — CHANGES_REQUESTED** [review]
Still some duplicated code, it may eventually change in different ways leading to hard-to-find bugs

File: src/trudp_ev.c [hunk 4]
```diff
@@ -84,7 +84,13 @@ void trudpSendQueueCbStart(trudpProcessSendQueueData *psd,
     }
 
     // If next_expected_time (net) or GetSendQueueTimeout
-    if((tt = (next_et != UINT64_MAX) ? next_et : trudpGetSendQueueTimeout(psd->td, ts)) != UINT32_MAX) {
```
Comment (kirill-scherba) [inline src/trudp_ev.c:87]: ???

**Review by kirill-scherba — CHANGES_REQUESTED** [review]

File: src/trudp_channel.c [hunk 5]
```diff
@@ -623,6 +630,40 @@ int trudpChannelProcessReceivedPacket(trudpChannelData *tcd, uint8_t *data,
...
+            uint64_t min_time = UINT64_MAX;
+            char *min_channel_key = NULL;
+            uint64_t current_time = teoGetTimestampFull();
+            teoMapIteratorReset(&it, tcd->td->map);
+            while((el = teoMapIteratorNext(&it))) {
+                trudpChannelData *tcd = (trudpChannelData *)teoMapIteratorElementData(el, NULL);
+                uint64_t expected_time = trudpSendQueueGetExpectedTime(tcd->sendQueue);
+                // if (expected_time == 0) {
+                //   continue;
+                // }
```
Comment (kshokhin) [inline src/trudp_channel.c]: Удалить это надо, чтобы не путало потом.

**Pull request closed without merging**